File: ValidPerfectSquare.java
class Solution {
    public boolean isPerfectSquare(int num) {
                if(num==1) return true;
          
        int n=num; int noOfDigits=0; int q=1;
        while(q>0){
            q=n/10;
            n=q;
            noOfDigits++;
        }
        long ans=num;
        for(int i=0;i<noOfDigits;i++){
            ans=ans/2;
        }
        for(long i=2;i<=ans;i++){
            if(i*i==num) return true;
            if(i*i>num) return false;
        }
        return false;

    }
}

File: ValidPerfectSquare.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        if(num==1) return true;
        
        int start=1; int end=num;
        while(start<=end){
        long long int mid=(end-start)/2 + start;
        if(mid*mid==num) return true;
        if(mid*mid>num) end=mid-1;
        if(mid*mid<num) start=mid+1;
        }
        return false;
    }
};